src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -84,6 +84,35 @@ enum AnchorFailure {
     RustdocAnchorConflict(Res),
 }
 
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+struct CacheKey {
+    module_id: DefId,
+    ns: Option<Namespace>,
```

Comment:
The namespace is not sufficient, you need the full disambiguator. For example, these are both in the type namespace but resolve to different things:

```rust
/// link to [primitive@char] and [mod@char]
pub mod char {}
```